src/main/java/weiweibot/storage/Storage.java

```diff
@@ -0,0 +1,158 @@
+     * @throws WeiExceptions on I/O errors or malformed records
+     */
+    public TaskList load() {
+        List<Task> list = new ArrayList<>();
+        if (!Files.exists(file)) {
+            return new TaskList(list);
+        }
+        try (BufferedReader br = Files.newBufferedReader(file)) {
+            String line;
+            int lineNo = 0;
+            while ((line = br.readLine()) != null) {
+                lineNo++;
+                String trimmed = line.trim();
+                if (trimmed.isEmpty()) {
+                    continue;
+                }
+                try {
+                    list.add(parseLine(trimmed));
+                } catch (RuntimeException ex) {
+                    throw new WeiExceptions("Corrupted line " + lineNo + ": " + trimmed);
+                }
+            }
+        } catch (IOException e) {
+            throw new WeiExceptions("Failed to load tasks: " + e.getMessage());
+        }
+        return new TaskList(list);
+    }
+
+    /**
+     * Saves all tasks to the backing file, creating parent directories if needed.
+     *
+     * @param tasks the tasks to persist
+     * @throws WeiExceptions on I/O errors
+     */
+    public void save(TaskList tasks) {
+        try {
+            Files.createDirectories(file.getParent() != null ? file.getParent() : Paths.get("."));
+            try (BufferedWriter bw = Files.newBufferedWriter(file)) {
+                for (Task t : tasks.asUnmodifiableList()) {
+                    bw.write(serialize(t));
+                    bw.newLine();
+                }
+            }
+        } catch (IOException e) {
+            throw new WeiExceptions("Failed to save tasks: " + e.getMessage());
+        }
+    }
+
+    private String serialize(Task t) {
+        String type;
+        String marked = t.isMarked() ? "1" : "0";
+        if (t instanceof Todo todo) {
+            type = "T";
+            return String.join(" | ", type, marked, todo.getDescription());
+        } else if (t instanceof Deadline d) {
+            type = "D";
+            String by = d.getBy().format(FILE_DT);
+            return String.join(" | ", type, marked, d.getDescription(), by);
+        } else if (t instanceof Event e) {
+            type = "E";
+            String from = e.getFrom().format(FILE_DT);
+            String to = e.getTo().format(FILE_DT);
+            return String.join(" | ", type, marked, e.getDescription(), from, to);
+        } else {
+            throw new WeiExceptions("Unknown task type for serialization: " + t.getClass());
+        }
+    }
+
+    private Task parseLine(String line) {
+        String[] parts = line.split("\\|");
+        List<String> tokens = new ArrayList<>();
+        for (String p : parts) {
+            tokens.add(p.trim());
+        }
+        if (tokens.size() < 3) {
+            throw new WeiExceptions("Invalid record: " + line);
+        }
+
+        String type = tokens.get(0);
+        boolean marked = switch (tokens.get(1)) {
```

Comment:
These inline switches would probably look more readable with a level of indentation for the cases